### ISSUE TITLE
Incorrect type checking (maybe?) for recursive types

### ISSUE THREAD

**Issue by e965** [microsoft/TypeScript#61520]
### 🔎 Search Terms

"recursive"

### 🕗 Version & Regression Information

Typescript 5.7.3, but on the current version in the sandbox (5.8.2) it reproduces too

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.8.2#code/C4TwDgpgBAShDGBXATgZwJYDcIBVzQF4oBvAKCigDN0IAbAEwH4AuKAO0QFsAjCZAbnJQA8jBawEKDNjyQA2gF1BAX0Gl4AezapgVdGmCsAFAEooBAHwSkaLLnzmop81aNkKFAPSeoqABYaiAxQvFB8yBrIVJFQwH7QqAha9AA0IYi6cXzQ6KjsGmHIEchCFKKscsR6dPSsAIxQymlVAIbc8KwATI0KKaUeA4OD3lAAemMT41OkyiZqmtq61AYAwpHICIZOZpbWUnayhNuscDbS9pAuTu4eI9wZUADu6HFhAB5gtOjwL1AbwCg2LEHPQELQWsgWsB0FooL94PF4ABrPKaIqbWggfrlKCVaoMeqNZpQNodKDdZS9fpDGm3HxTSaTGZzIA

### 💻 Code

```ts
type RecursiveType = {
  field?: number;
  OR?: RecursiveType[];
};

const first: () => RecursiveType = () => ({
    // should be error for the second, but there is no error
    OR: [{ field: 1 }, { abc: 2 }],
                    // ^ ^ ^ ^ ^ ^
});

const firstCorrect: () => RecursiveType = (): RecursiveType => ({
    // but with explicit return type declaration it checks correctly
    OR: [{ field: 1 }, { abc: 2 }],
                    // ^ ^ ^ ^ ^ ^
});
```

### 🙁 Actual behavior

Only the second example throws an error

### 🙂 Expected behavior

Both examples should throw an error

### Additional information about the issue

This is probably ACTUALLY correct behavior, but I still decided to create a bug report just in case

**Comment by MartinJohns** [comment]
Working as intended. The return type to your function is inferred based on the returned value, not on the type of the variable you assign it to. Then on the second round it checks assignability, and the types are compatible so there's no error. It's allowed to have additional properties.

There's a bunch of duplicates of this on the tracker.

**Comment by Andarist** [comment]
You are totally right that this is a fallout of how things work currently. But this issue here is quite specific so I wouldn't call it a hard duplicate of the referenced issue (yet?). 

The problem with this is that the return type "suffers" from object literal unification~ (introduced in [TS 2.8](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#improved-type-inference-for-object-literals)) but at the same time, the contextual type for that array element is a weak type. Even though extra properties are allowed there (as explained by you) - that type is only compatible with the weak type target because of that object literal unification. Without it, it wouldn't be.

**Comment by e965** [comment]
The funniest thing for me is that exactly the same .d.ts is generated for both expressions (as it should be), but the behavior is different

![Image](https://github.com/user-attachments/assets/c6b28a42-ed26-493e-8872-71721dd1fb85)

**Comment by nmain** [comment]
> but the behavior is different

From a typing perspective, it's not.  Any object type can always have excess properties on it at runtime.  It'd be nice to have the excess property checks lint here, but it's still just a lint.
